### ISSUE TITLE
Licence

### ISSUE THREAD

**Issue by Apskaita5** [MarimerLLC/cslaforum#826]
Is it allowed to use CSLA code to create an alternative open source framework with reduced functionality?
I'm using an old version of CSLA and only need a few new features. At the moment CSLA has grown  too large for me. It would be more simple for me to make my own version by only taking few CLSA components.

**Comment by ajj7060** [comment]
Just curious, what do you mean by too big?  Size?  Amount of code you need to write?

**Comment by Apskaita5** [comment]
Amount of code and functionality within the CSLA framework itself.
I find it too difficult to understand internal mechanisms and I don't like to use something that I don't understand to the last detail. The framework got too intrusive as well. I like when a framework helps to do trivial repeating tasks but not limit myself, e.g. like property getters and setters.
Basically I need version 2.0 rewritten for .NET standard 2 plus async support and validation attributes.

**Comment by rockfordlhotka** [comment]
@Apskaita5 

For private purposes inside an org you can pretty much do what you'd like. The older versions of CSLA have a license the prevents you from making money directly off CSLA or derivatives.

Which specific version do you want to fork? That'll determine the license we're talking about.

**Comment by rockfordlhotka** [comment]
Not to be negative, but I do need to point out the relative absurdity of this statement

>  I don't like to use something that I don't understand to the last detail

You use .NET, very possibly Newtonsoft.Json, Entity Framework, XAML, Razor, etc. Those are FAR more complex than anything CSLA does. So if you understand all of those open source frameworks to the last detail it is hard to imagine you can't understand CSLA to the last detail.

😄

**Comment by Chicagoan2016** [comment]
@Apskaita5, not trying to steal your thread but I have a question for @rockfordlhotka 

@rockfordlhotka , did you ever think of making a commercial version of Csla? I for sure would buy it in a heart beat. One of your close friends from California used to sell a commercial framework which I happened to use in the past and liked it.

Kind regards

**Comment by Apskaita5** [comment]
I'd like to fork the current version, yet cut out around 3/4 of the code. There won't be much code borrowed, it's more like the base architecture: rule engine (modified not to use custom property management + no state modification); most of data portal functionality;  maybe mobile formatter, though i also consider proxy model, because i use dataportal as "dynamic rest"; that's pretty much all.

**Comment by rockfordlhotka** [comment]
@Apskaita5 The current code is under the MIT license, which is extremely permissive.

I personally think creating a fork as you suggest is not a good plan, but the license does permit it.

**Comment by rockfordlhotka** [comment]
@Chicagoan2016 yes, about 12 years ago I looked very closely at creating a commercial version of CSLA. Ultimately decided against it, but did give it a great deal of thought.

**Comment by Apskaita5** [comment]
Ok, let me clarify :)
Programming is my hobby, not a profession. Therefore, i can choose what i like to develop, which happens to be: financial accounting applications, CRM's and DMS's. (the last one is irrelevant for csla)
I love business modelling part. Presentation, persistence are required, but i only do that to make the business layer available for the end users. 
Yet for the business layer i want to have full control and understanding of everything that happens with business data in between of persistence and presentation.
I don't criticize your work. Just my personal preferences.
CSLA is used in my open source financial accounting application that is within top 5 in Lithuania. And it does its job without a single bug in 5+ years for hundreds of users. (Lithuania is a small country, not that many accountants arround) All i need (would like to) is a "light version".
The result would be open source. Yet sometimes i do commercial projects and would also like to use it.

**Comment by rockfordlhotka** [comment]
@Apskaita5 no problem, you can fork the repo and create your own version.

My personal preference would be that you put that energy into helping to improve CSLA, possibly making things you see as too complex be easier to use.

But as long as you name your fork something other than CSLA you can go ahead with your plan.

**Comment by Apskaita5** [comment]
Shall I add some copyright notices for files from the CSLA framework?
Some concepts in the current version (like compulsory property management) is just too much for my religious feelings. Sorry.

**Comment by JasonBock** [comment]
Here's my $0.02....

There's a legal aspect to the question, "Is it allowed to use CSLA code to create an alternative open source framework with reduced functionality?" That I'm not going to answer, and @rockfordlhotka has already addressed it anyway.

But there's the other aspect that is mentioned: that CSLA is "heavy", a "lightweight" or "streamlined" version is desired, etc. I've used CSLA for the last 10 years and I've seen clients take CSLA and make their own version. Every single time, it's never ended well, and it really has nothing to do with CSLA per-se. Take _any_ popular package you can find in NuGet that's not trivial (e.g. Benchmark.NET), and there's part to it you will never use. However, trying to maintain your own version is never easy. I'm not saying people aren't intelligent enough to understand how these frameworks work; it's a matter of time and effort. I've seen clients take a stripped-down version of CSLA and end up in a position where no one maintains what they have, or they end up implementing what they removed in the first place.

I'm not trying to convince the original author of this thread to not go down the path they're convinced they should go down. I'm just offering my advice to go down a different path. Work with the CSLA community to make the product better by contributing to the repo. Then others gain from your work. I've done this myself by making some improvements in the performance of the serializer and by writing the analyzers in CSLA. I encourage others to do the same.

**Comment by Chicagoan2016** [comment]
> Here's my $0.02....
> 
> There's a legal aspect to the question, "Is it allowed to use CSLA code to create an alternative open source framework with reduced functionality?" That I'm not going to answer, and @rockfordlhotka has already addressed it anyway.
> 
> But there's the other aspect that is mentioned: that CSLA is "heavy", a "lightweight" or "streamlined" version is desired, etc. I've used CSLA for the last 10 years and I've seen clients take CSLA and make their own version. Every single time, it's never ended well, and it really has nothing to do with CSLA per-se. Take _any_ popular package you can find in NuGet that's not trivial (e.g. Benchmark.NET), and there are parts to it you will never use. However, trying to maintain your own version is never easy. I'm not saying people aren't intelligent enough to understand how these frameworks work; it's a matter of time and effort. I've seen clients take a stripped-down version of CSLA and end up in a position where no one maintains what they have, or they end up implementing what they removed in the first place.
> 
> I'm not trying to convince the original author of this thread to not go down the path they're convinced they should go down. I'm just offering my advice to go down a different path. Work with the CSLA community to make the product better by contributing to the repo. Then others gain from your work. I've done this myself by making some improvements in the performance of the serializer and by writing the analyzers in CSLA. I encourage others to do the same.

@JasonBock , thanks for your input, I really like what you said, looking at your profile I have almost the educational background as you but I wish I was 10% as talented as you are.
have a good day

Regards

**Comment by Chicagoan2016** [comment]
> @Chicagoan2016 yes, about 12 years ago I looked very closely at creating a commercial version of CSLA. Ultimately decided against it, but did give it a great deal of thought.

@rockfordlhotka , considering myself as your student I still think you should make a commercial version of Csla : )
You sir have provided a great service to .NET community for as long as I could remember, growing up in the 80s and watching movies like 'NightShift' (Micheal Keaton) I think there is nothing wrong with earning some real 'do-re-mi' :) :).
Best Regards
